addons/controllers/clusterbootstrap_controller.go

```diff
@@ -606,26 +994,47 @@ func (r *ClusterBootstrapReconciler) updateValuesFromProvider(cluster *clusterap
 		providerLabels := newProvider.GetLabels()
 		if providerLabels == nil {
 			newProvider.SetLabels(map[string]string{
-				types.PackageNameLabel: pkg.RefName,
+				// A package refName could contain characters that K8S does not like as a label value.
+				// For example, kapp-controller.tanzu.vmware.com.0.30.0+vmware.1-tkg.1 is a
+				// valid package refName, but it contains "+" that K8S complains. We parse the refName by replacing
+				// + to ---.
+				types.PackageNameLabel: util.ParseStringForLabel(pkg.RefName),
 				types.ClusterNameLabel: cluster.Name,
 			})
 		} else {
-			providerLabels[types.PackageNameLabel] = pkg.RefName
+			providerLabels[types.PackageNameLabel] = util.ParseStringForLabel(pkg.RefName)
 			providerLabels[types.ClusterNameLabel] = cluster.Name
 		}
 
-		newProvider.SetName(fmt.Sprintf("%s-%s-package", cluster.Name, packageShortName(pkg.RefName)))
-		log.Info("cloning provider", "provider", newProvider)
-		newProvider, err = r.dynamicClient.Resource(*gvr).Namespace(bootstrap.Namespace).Create(r.context, newProvider, metav1.CreateOptions{})
+		newProvider.SetName(fmt.Sprintf("%s-%s-package", cluster.Name, pkgRefName))
+		newProvider.SetNamespace(bootstrap.Namespace)
+		log.Info(fmt.Sprintf("cloning provider %s/%s to namespace %s", cbTemplateNamespace, newProvider.GetName(), bootstrap.Namespace), "gvr", gvr)
+		// newProvider and createdOrUpdatedProvider are different. The newProvider is the one we want apiserver to accept,
+		// however createdOrUpdatedProvider is the actual object pointer that apiserver creates with several managed fields,
+		// i.e., resourceVersion. When error happens on create or update, createdOrUpdatedProvider will be nil, but newProvider
+		// won't. The intent of this function is to return the actual object pointer that apiserver creates, so we should
+		// return createdOrUpdatedProvider instead of newProvider. Otherwise, when the caller wants to make changes to the
+		// created provider objects, there will errors. I.e., [invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update]
+		createdOrUpdatedProvider, err = r.dynamicClient.Resource(*gvr).Namespace(bootstrap.Namespace).Create(r.context, newProvider, metav1.CreateOptions{})
 		if err != nil {
-			log.Error(err, "unable to clone provider", "provider", newProvider, "gvr", gvr)
+			// There are possibilities that current reconciliation loop fails due to various reasons, and during next reconciliation
+			// loop, it is possible that the provider resource has been created. In this case, we want to run update/patch.
+			if apierrors.IsAlreadyExists(err) {
+				createdOrUpdatedProvider, err = r.dynamicClient.Resource(*gvr).Namespace(bootstrap.Namespace).Update(r.context, newProvider, metav1.UpdateOptions{})
+				if err != nil {
+					log.Info(fmt.Sprintf("updated provider %s/%s", createdOrUpdatedProvider.GetNamespace(), createdOrUpdatedProvider.GetName()), "gvr", gvr)
+					return newProvider, nil
```

Comment:
Also in that case when the update fails, ```createdOrUpdatedProvider``` would be nil and seems like you needed to record the name and the ```createdOrUpdatedProvider.GetName()``` & ```createdOrUpdatedProvider.GetName()``` into another variables after resource Get; as their access here can end up in nil pointer access for when the update fails

Reply:
And your log message here needs to get improved anyway, as apparently the update has failed here, not succeeded